### PR TITLE
[Feature](bangc-ops): add new op roipoint_pool3d.

### PR DESCRIPTION
**Pull request by DanieeelLiu**
Thanks for your contribution and we appreciate it a lot. 

## 1. Motivation

Please describe your motivation and the goal you want to achieve through this pull request.

## 2. Modification

modified:   bangc-ops/kernel_depends.toml
new file:   bangc-ops/kernels/roipoint_pool3d/roipoint_pool3d.cpp
new file:   bangc-ops/kernels/roipoint_pool3d/roipoint_pool3d_union1.mlu
new file:   bangc-ops/kernels/roipoint_pool3d/roipoint_pool3d_union1_large_boxes_num.mlu
modified:   bangc-ops/mlu_op.h
modified:   bangc-ops/mlu_op_kernel.h
modified:   bangc-ops/test/mlu_op_gtest/pb_gtest/mlu_op_test_proto/mlu_op_test.proto
new file:   bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/roipoint_pool3d/roipoint_pool3d.cpp
new file:   bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/roipoint_pool3d/roipoint_pool3d.h
new file:   bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/roipoint_pool3d/test_case/case_0.prototxt
new file:   docs/bangc-docs/design_docs/roipoint_pool3d/nram_roipoint_pool3d.png
new file:   docs/bangc-docs/design_docs/roipoint_pool3d/roipoint_pool3d.md
new file:   docs/bangc-docs/design_docs/roipoint_pool3d/test_element_zero.png

## 3. Test Report

If you want to know how to do operator testing, you can see [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).

#### 3.1.1 Accuracy Acceptance Standard

For static threshold standard details, see: [MLU-OPS Accuracy Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Accuracy-Acceptance-Standard.md).

- [x] diff1: diff1 <= 3e-3
- [x] diff2: diff2 <= 3e-3

#### 3.1.2 Operator Scheme checklist

|     No.        |                 Details              |            Check Results             |
|----------------|--------------------------------------|--------------------------------------|
|        1       |Supported hardware                    |             MLU370<br>MLU590         |
|        2       |Job types                             |          block <br> U1 <br> U4       |
|        3       |Layouts                               |          NHWC 、NCHW、ARRAY etc      |
|        4       |Whether multi-dimensions are supported|                                      |
|        5       |Whether element zero is supported     |                                      |
|        6       |Data type(half/float)                 |           half / float etc           |
|        7       |Whether there is size limit           |                                      |

#### 3.1.3 New Feature Test

If you have checked the following items, please tick the relevant box.

- [ ] Data type test
- [ ] Multidimensional tensor test
- [ ] Layout test
- [ ] Different size/integer remainder end segment/alignment misalignment test
- [ ] Zero dimensional tensor test/zero element test
- [ ] stability test
- [ ] Multiple platform test
- [ ] Gen_case module test
- [ ] Nan/INF tests 
- [ ] Bug fix tests
- [ ] For memory leak check details, see[GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).
- [ ] For code coverage check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).
- [ ] For I/O calculation efficiency check details, see: [MLU-OPS Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md).

#### 3.1.4 Parameter Check

When a new operator is submitted, the test points are given and the test results are stated.

|                   Test Point                    | Acceptance Standard | Test Result (Error Message) |
| ----------------------------------------------- | --------------------| --------------------------- |
| Whether it conforms to the operator restriction |     Normal error    |                             |
| Whether illegal parameters are passed           |     Normal error    |                             |

### 3.2 Accuracy Test

For the cases used in the New Feature Test section, the features and the number of cases are recorded here. When multiple operations are tested, multiple tables are needed to include details of these operations.

Operation:

|Test Point           | Description                      | Quantity |  Comment |
|----------           |----------------------------------|----------|  --------|
|Data type test       |half/float/int8                   |          |          |
|Mult-tensor test     |Supports 1-8 dims                 |          |          |
|Layout test          |Supports NCHW/NHWC                |          |          |
|Zero element test    |Whether to support this test      |          |          |
|Stability test       |--gtest_repeat=NUM<br>--thread=NUM|          |          |
|Mult-platform test   |MLU370/MLU590                     |          |          |
|Nan/INF test         |Whether to support this test      |          |          |

### 3.3 Performance Test

See [MLU-OPS Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md) for details.

Platform：MLU370

|Operation|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-------|----|----|-----|----|----|----|-----|
|op_name|    |    |     |    |    |    |     |
|op_name|    |    |     |    |    |    |     |

Platform：MLU590

|Operation|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-------|----|----|----|----|----|----|-----|
|op_name|    |    |    |    |    |    |     |
|op_name|    |    |    |    |    |    |     |

### 3.4 Summary Analysis

370: release_temp(72) + release test(16)
[----------] Global test environment tear-down
[ SUMMARY  ] Total 88 cases of 1 op(s).
ALL PASSED.
[==========] 88 test cases from 1 test suite ran. (123717 ms total)
[  PASSED  ] 88 test cases.

590: release_temp(72) + release test(16)
[----------] Global test environment tear-down
[ SUMMARY  ] Total 88 cases of 1 op(s).
ALL PASSED.
[==========] 88 test cases from 1 test suite ran. (173275 ms total)
[  PASSED  ] 88 test cases.

### CODE REVIEW

File: bangc-ops/mlu_op.h [hunk 1]
```diff
@@ -8478,6 +8478,177 @@ mluOpIndiceConvolutionBackwardFilter(mluOpHandle_t handle,
...
+ * see ::mluOpTensorDescriptor_t.
+ * @param[out] size
+ * A host pointer to the returned size of the extra workspace in bytes that is
+ * used in the ::mluOpRoiPointPool3d operation.
+ * @par Return
+ * - ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_BAD_PARAM
+ *
+ * @par API Dependency
+ * - The allocated extra workspace should be passed to the ::mluOpRoiPointPool3d function
+ *   to perform the ::mluOpRoiPointPool3d operation.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:8527]: 去掉function后面的半句话。

File: bangc-ops/mlu_op.h [hunk 2]
```diff
@@ -8478,6 +8478,177 @@ mluOpIndiceConvolutionBackwardFilter(mluOpHandle_t handle,
...
+ *
+ * @par Data Type
+ * - The supported data types for input and output are as follows:
+ *   <b>Note that the data type of \b points, \b point_features, \b boxes3d, and
+ *   \b pooled_features must be the same.
+ *   - \b points: half, float
+ *   - \b point_features: half, float
+ *   - \b boxes3d: half, float
+ *   - \b pooled_features: half, float
+ *   - \b pooled_empty_flag: int32
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:8614]: 这5个前面的\b全部去掉

File: docs/bangc-docs/user_guide/9_operators/index.rst [hunk 3]
```diff
@@ -425,4 +425,17 @@ CarafeForward的反向功能，即根据输入特征图、上采样核函数的
...
 ----------------------------------
-一种通用、轻量且非常有效的上采样算法，在物体识别、语义分割、图像修复等任务上都展示出很好的效果。
\ No newline at end of file
+一种通用、轻量且非常有效的上采样算法，在物体识别、语义分割、图像修复等任务上都展示出很好的效果。
+
+.. _roipoint_pool3d:
+
+mluOpRoiPointPool3d
+----------------------------------
+该算子功能是筛选出3D bounding boxes内的点云数据坐标和特征。LiDAR坐标系下，判断点云数据坐标是否在box边框内的计算公式为：
```
Comment (GraceGuanH) [inline docs/bangc-docs/user_guide/9_operators/index.rst:434]: 是否在bounding box边框内的计算公式为：

File: docs/bangc-docs/user_guide/9_operators/index.rst [hunk 4]
```diff
@@ -425,4 +425,17 @@ CarafeForward的反向功能，即根据输入特征图、上采样核函数的
 
 mluOpCarafeForward
 ----------------------------------
-一种通用、轻量且非常有效的上采样算法，在物体识别、语义分割、图像修复等任务上都展示出很好的效果。
\ No newline at end of file
+一种通用、轻量且非常有效的上采样算法，在物体识别、语义分割、图像修复等任务上都展示出很好的效果。
+
+.. _roipoint_pool3d:
```
Comment (GraceGuanH) [inline docs/bangc-docs/user_guide/9_operators/index.rst:430]: 这次新增了两个接口，mluOpGetRoiPointPool3dWorkspaceSize也需要加到这里

**Review by GraceGuanH — COMMENTED** [review]
Please check the comments.

File: bangc-ops/mlu_op.h (hunk 1 above)
Comment (DanieeelLiu) [reply bangc-ops/mlu_op.h]: 已修改

File: bangc-ops/mlu_op.h (hunk 2 above)
Comment (DanieeelLiu) [reply bangc-ops/mlu_op.h]: 已修改

File: docs/bangc-docs/user_guide/9_operators/index.rst (hunk 3 above)
Comment (DanieeelLiu) [reply docs/bangc-docs/user_guide/9_operators/index.rst]: 已修改

File: docs/bangc-docs/user_guide/9_operators/index.rst (hunk 4 above)
Comment (DanieeelLiu) [reply docs/bangc-docs/user_guide/9_operators/index.rst:660]: workspace 接口不添加

File: docs/bangc-docs/design_docs/roipoint_pool3d/roipoint_pool3d.md [hunk 5]
```diff
@@ -0,0 +1,553 @@
...
+                                             size_t workspace_size,
+                                             const mluOpTensorDescriptor_t pooled_features_desc,
+                                             void *pooled_features,
+                                             const mluOpTensorDescriptor_t pooled_empty_flag_desc,
+                                             void *pooled_empty_flag);
+
+mluOpStatus_t MLUOP_WIN_API mluOpGetRoiPointPool3dWorkspaceSize(mluOpHandle_t handle,
+                                                             const int batch_size,
+                                                             const int pts_num,
+                                                             const int boxes_num,
```
Comment (WangYuan-cn) [inline docs/bangc-docs/design_docs/roipoint_pool3d/roipoint_pool3d.md:189]: 两个接口错位了

File: bangc-ops/mlu_op.h [hunk 6]
```diff
@@ -8669,6 +8669,176 @@ mluOpIndiceConvolutionBackwardFilter(mluOpHandle_t handle,
...
+ * - The allocated extra workspace should be passed to the ::mluOpRoiPointPool3d.
+ *
+ * @note
+ * - None.
+ *
+ * @par Requirements
+ * - None.
+ *
+ * @par Example
+ * - None.
```
Comment (WangYuan-cn) [inline bangc-ops/mlu_op.h:8726]: 增加reference

File: bangc-ops/kernels/roipoint_pool3d/roipoint_pool3d.cpp [hunk 7]
```diff
@@ -0,0 +1,396 @@
...
+#include "core/tensor.h"
+#include "core/type.h"
+#include "core/tool.h"
+#include "kernels/kernel.h"
+#include "mlu_op.h"
+#include "mlu_op_kernel.h"
+
+#define MAX(a, b) ((a) > (b) ? (a) : (b))
+
+mluOpStatus_t paramcheck(const int batch_size, const int pts_num,
```
Comment (WangYuan-cn) [inline bangc-ops/kernels/roipoint_pool3d/roipoint_pool3d.cpp:36]: static

File: docs/bangc-docs/design_docs/roipoint_pool3d/roipoint_pool3d.md (hunk 5 above)
Comment (DanieeelLiu) [reply docs/bangc-docs/design_docs/roipoint_pool3d/roipoint_pool3d.md]: Fixed

File: bangc-ops/mlu_op.h (hunk 6 above)
Comment (DanieeelLiu) [reply bangc-ops/mlu_op.h:8732]: Fixed

File: bangc-ops/kernels/roipoint_pool3d/roipoint_pool3d.cpp (hunk 7 above)
Comment (DanieeelLiu) [reply bangc-ops/kernels/roipoint_pool3d/roipoint_pool3d.cpp]: Fixed

**Review by WangYuan-cn — APPROVED** [review]

File: bangc-ops/test/mlu_op_gtest/pb_gtest/mlu_op_test_proto/mlu_op_test.proto [hunk 8]
```diff
@@ -552,6 +553,11 @@ message RoiawarePool3dBackwardParam {
   optional int32 max_pts_each_voxel = 7 [default = 128];
 }
 
+// para to call mluOpRoiPointPool3d
```
Comment (duzekunKTH) [inline bangc-ops/test/mlu_op_gtest/pb_gtest/mlu_op_test_proto/mlu_op_test.proto:556]:
```suggestion
// param to call mluOpRoiPointPool3d
```

Comment (DanieeelLiu) [reply bangc-ops/test/mlu_op_gtest/pb_gtest/mlu_op_test_proto/mlu_op_test.proto]: Fixed

**Pull request merged**